### ISSUE TITLE
General Guidelines for Scaling PDE Residuals and Weighting Hard Constraints in Augmented Lagrangian PINNs

### ISSUE THREAD

**Issue by maxnleo** [lululxvi/deepxde#1947]
First of all, thank you for providing this hPINN code and for sharing your research---it's been immensely helpful in my work.

I have already reviewed the FAQ discussions related to 
\emph{``I failed to train the network or get the right solution, e.g., large training loss, unbalanced losses''}, but I still have further questions:

In addition to my earlier question about specific scaling factors \(\bigl(0.01, \, 10^2, \, \mu_{\mathrm{PDE}} = 0.1, \, \mu_{V} = 10^4\bigr)\), I'm also curious about \textbf{general best practices} for scaling PDE residuals and weighting constraints using an augmented Lagrangian method in PINNs. Specifically:

\begin{enumerate}
  \item \textbf{Balancing PDE Residuals vs.\ Constraints}\\
    Do you typically begin by normalizing the PDE residuals before adding the inequality-constrained terms in the loss? Any rules of thumb for how large the multipliers on the PDE terms should be compared to the multiplier on the inequality constraint (e.g.\ \(\mu_{V}\) for the volume constraint)?

  \item \textbf{Choosing Penalty Parameters}\\
    Besides trial-and-error, are there recommended heuristics (perhaps involving dimensionless groups like the Reynolds number \(\mathrm{Re}\) or the ratio \(\alpha_{\max}/\alpha_{\min}\) in Brinkman penalization) for selecting \(\mu_{\mathrm{PDE}}\), \(\mu_{V}\), or the steps for updating \(\lambda\)? Have you tried an adaptive or staged approach (e.g., gradually increasing these parameters) to reduce instabilities?

  \item \textbf{Avoiding Numerical Instabilities}\\
    If we scale the PDE residual incorrectly (too large or too small), are there typical signs in training (e.g.\ oscillatory or exploding losses) that indicate an imbalance in the loss terms? What diagnostic methods would you recommend?
\end{enumerate}

Any additional insights or principles on systematically tuning or scaling PDE residuals, along with augmented Lagrangian parameters in PINNs, would be greatly appreciated. Thank you again for your time and for making your work so accessible!

**Comment by lululxvi** [comment]
The basic rule is to choose the weights to balance all losses. Dimensionless is usually needed.
